mig-scheduler/agents_queues.go

```diff
@@ -6,167 +6,13 @@
-	if err != nil {
-		panic(err)
-	}
-
-	err = ctx.MQ.Chan.Qos(0, 0, false)
-	if err != nil {
-		panic(err)
-	}
-
-	heartbeatChan, err = ctx.MQ.Chan.Consume(mig.QueueAgentHeartbeat, "", true, false, false, false, nil)
-	if err != nil {
-		panic(err)
-	}
-	ctx.Channels.Log <- mig.Log{OpID: ctx.OpID, Desc: "agents heartbeats listener initialized"}
-
-	return
-}
-
-// getHeartbeats processes the heartbeat messages sent by agents
-func getHeartbeats(msg amqp.Delivery, ctx Context) (err error) {
-	defer func() {
-		if e := recover(); e != nil {
-			err = fmt.Errorf("getHeartbeats() -> %v", e)
-		}
-		if ctx.Debug.Heartbeats {
-			ctx.Channels.Log <- mig.Log{OpID: ctx.OpID, Desc: "leaving getHeartbeats()"}.Debug()
-		}
-	}()
-	// a normal heartbeat should be between 500 and 5000 characters, some may be larger if
-	// large environments are collected, so we fix the upper limit to 100kB
-	if len(msg.Body) > 102400 {
-		panic("discarded heartbeat larger than 100kB")
-	}
-	var agt mig.Agent
-	err = json.Unmarshal(msg.Body, &agt)
-	if err != nil {
-		panic(err)
-	}
-	if ctx.Debug.Heartbeats {
-		desc := fmt.Sprintf("Received heartbeat for Agent '%s' QueueLoc '%s'", agt.Name, agt.QueueLoc)
-		ctx.Channels.Log <- mig.Log{Desc: desc}.Debug()
-	}
-	// discard expired heartbeats
-	agtTimeOut, err := time.ParseDuration(ctx.Agent.TimeOut)
-	if err != nil {
-		panic(err)
-	}
-	expirationDate := time.Now().Add(-agtTimeOut)
-	if agt.HeartBeatTS.Before(expirationDate) {
-		desc := fmt.Sprintf("Expired heartbeat received from Agent '%s'", agt.Name)
-		ctx.Channels.Log <- mig.Log{Desc: desc}.Notice()
-		return
-	}
-	// replace the heartbeat with current time
-	agt.HeartBeatTS = time.Now()
-	// do some sanity checking
-	if agt.Mode != "" && agt.Mode != "daemon" && agt.Mode != "checkin" {
-		panic(fmt.Sprintf("invalid mode '%s' received from agent '%s'", agt.Mode, agt.QueueLoc))
-	}
-	if len(agt.Name) > 1024 {
-		panic(fmt.Sprintf("agent name longer than 1024 characters: name '%s' from '%s'", agt.Name, agt.QueueLoc))
-	}
-	if len(agt.Version) > 128 {
-		panic(fmt.Sprintf("agent version longer than 128 characters: version '%s' from '%s'", agt.Version, agt.QueueLoc))
-	}
-	// if agent is not authorized, ack the message and skip the registration
-	// nothing is returned to the agent. it's simply ignored.
-	ok, err := isAgentAuthorized(agt.QueueLoc, ctx)
-	if err != nil {
-		panic(err)
-	}
-	if !ok {
-		desc := fmt.Sprintf("getHeartbeats(): Agent '%s' is not authorized", agt.QueueLoc)
-		ctx.Channels.Log <- mig.Log{Desc: desc}.Warning()
-		// agent authorization failed so we drop this heartbeat and return
-		return
-	}
-
-	// write to database in a goroutine to avoid blocking
-	go func() {
-		// if an agent already exists in database, we update it, otherwise we insert it
-		agent, err := ctx.DB.AgentByQueueAndPID(agt.QueueLoc, agt.PID)
-		if err != nil {
-			agt.DestructionTime = time.Date(9998, time.January, 11, 11, 11, 11, 11, time.UTC)
```

Comment:
I looked into why the tests are failing, and it turns out some of the fields in the db are no longer getting added:
```
destructiontime | 0001-01-01 00:00:00+00
heartbeattime   | 0001-01-01 00:00:00+00
refreshtime     | 0001-01-01 00:00:00+00
```
I looked at what a good agent document looks like, and it's like so:
```
destructiontime | 9998-01-11 11:11:11+00
heartbeattime   | 2018-11-12 21:30:36.123557+00
refreshtime     | 2018-11-12 21:28:06.215546+00
```

I think we lost some of the logic here when moving this to the API FYI @zsck

Reply:
The status field is also not set, when in the other one, it's set to 'online'. I think we're missing some function here to identify which agents are online and which are offline.

Reply:
Good callout.  I didn't think it was going to matter. In fact, I'm not sure our tests are failing because of this.  I've pushed a fix to include these fields and we still see things failing in Travis.

Reply:
Looks to me like the tests might be failing due to them being run [before Postgres is fully initialized](https://travis-ci.org/mozilla/mig/builds/454728538#L2095). That's my hypothesis based on this error, anyway.

Reply:
I forgot a key part, when I was doing my testing, I had merged https://github.com/mozilla/mig/pull/467 into my local branch, so my test failures were different than what travisci was reporting.